backend-rust/src/graphql_api.rs

```diff
@@ -1105,22 +1104,127 @@ impl Account {
+            sqlx::query_scalar!("SELECT COUNT(*) FROM transactions WHERE sender=$1", self.index)
+                .fetch_one(get_pool(ctx)?)
+                .await?;
+        Ok(count.unwrap_or(0))
     }
 
     async fn tokens(
         &self,
-        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: i32,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: Option<u64>,
         #[graphql(desc = "Returns the elements in the list that come after the specified cursor.")]
-        after: String,
-        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: i32,
+        after: Option<String>,
+        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: Option<u64>,
         #[graphql(desc = "Returns the elements in the list that come before the specified cursor.")]
-        before: String,
+        before: Option<String>,
     ) -> ApiResult<connection::Connection<String, AccountToken>> {
-        todo_api!()
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+        let query = ConnectionQuery::<i64>::new(
+            first,
+            after,
+            last,
+            before,
+            config.contract_connection_limit,
+        )?;
+
+        // Tokens with 0 balance are filtered out. We still display tokens with a
+        // negative balance (buggy cis2 smart contract) to help smart contract
+        // developers to debug their smart contracts.
+        // Excluding tokens with 0 balances does not scale well currently. We might need
+        // to introduce a specific index to optimize this query in the future.
+        let mut row_stream = sqlx::query_as!(
+            AccountTokenInterim,
+            "
+            SELECT 
+                token_id,
+                contract_index,
+                contract_sub_index,
+                raw_balance,
+                account_id,
+                row_num
+            FROM (
+                SELECT 
+                    token_id,
+                    contract_index,
+                    contract_sub_index,
+                    raw_balance,
+                    account_id,
+                    row_num
+                FROM (
+                    SELECT
+                        token_id,
+                        contract_index,
+                        contract_sub_index,
+                        balance AS raw_balance,
+                        account_index AS account_id,
+                        ROW_NUMBER() OVER (ORDER BY account_tokens.index) - 1 AS row_num
+                    FROM account_tokens
+                    JOIN tokens
+                        ON tokens.index = account_tokens.token_index
+                    WHERE account_tokens.balance != 0 
+                        AND account_tokens.account_index = $5
+                ) AS filtered_tokens
+                WHERE $1 < row_num AND row_num < $2
+                ORDER BY
```

Comment:
We order by the row number here, but these were not assigned in any particular order, so it seems a bit random what order we get in the end.
Do you know the order of the .NET backend?

Reply:
https://github.com/Concordium/concordium-scan/blob/main/backend/Application/Api/GraphQL/Accounts/Account.cs#L84